### ISSUE TITLE
Output do not update after close and reopen the notebook page

### ISSUE THREAD

**Issue by oeway** [jupyter/jupyter#83]
Hi,

I am using a library named keras with jupyter notebook. It's nice that keras use a text progress bar to indicate the progress like this:

```
Train on 381734 samples, validate on 20092 samples
Epoch 1/1
381734/381734 [==============================] - 257s - loss: 0.4094 - val_loss: 0.2863
```

But the problem is that, if the script is running with the progress bar, I can't close the webpage which I used to start the script. Otherwise I won't see the progress bar update anymore, when I open the notebook again, if it's still running, the title will become something like "(Starting) XXXXXX", but I can't see any update in the page.

The progress bar used by keras is like this: 
https://github.com/fchollet/keras/blob/b126b6328a44fa3332d2d7fd011da3ff196a669a/keras/utils/generic_utils.py

Can anyone look into this?

Thanks in advance.

**Comment by willingc** [comment]
Closing this issue as it's been inactive for several months. Please feel free to open a new issue at the [jupyter/help repo](https://github.com/jupyter/help) or reopen this issue. Thanks! :sunflower:

**Comment by hadim** [comment]
I am facing the same issue. Is it possible to reopen a notebook and get the output of the running kernel ?

**Comment by raviqqe** [comment]
I have the same issue too. Reconnecting kernel doesn't help me either.

**Comment by willingc** [comment]
@raviqqe Does this happen with Keras or any notebook?

**Comment by raviqqe** [comment]
I was running not Keras but numpy code. I closed a notebook tab and opened it again. However, it's not showing updated output while its kernel is still running. Note that I'm running Jupyter in client-server setting.

My environment is the below.
- Jupyter: 4.2.0
- Browser: Google Chrome 53.0.2785.154
- Host OS: Ubuntu 16.04.1 LTS (Xenial Xerus)
- Client OS: Chrome OS

The early part of my configuration file `~/.jupyter/jupyter_notebook_config.py`:
```
# Configuration file for jupyter-notebook.

#------------------------------------------------------------------------------
# Application(SingletonConfigurable) configuration
#------------------------------------------------------------------------------

## This is an application.

c = get_config()
c.NotebookApp.open_browser = False
c.NotebookApp.password = 'sha1:6cea77f:673802e58d129779298e501a2bc40'
c.NotebookApp.certfile = '/home/raviqqe/mycert.pem'
c.NotebookApp.keyfile = '/home/raviqqe/mykey.key'
c.NotebookApp.ip = '*'
c.NotebookApp.port = 8888

## The date format used by logging formatters for %(asctime)s
#c.Application.log_datefmt = '%Y-%m-%d %H:%M:%S'
...
```

I'm not sure that this problem is caused by Jupyter or Chrome.

**Comment by soerendip** [comment]
Same problem

**Comment by floatdrop** [comment]
@willingc can you reopen this issue? It hit me as well.

**Comment by willingc** [comment]
@floatdrop Instead of reopening here, would you mind opening a new issue on `jupyter/notebook` repo? Also Jupyter notebook has a new (2 days ago) 5.0 release. Perhaps upgrade and see if the issue is still present. Thanks!

**Comment by tadas-subonis** [comment]
The issue is still here

**Comment by ajinkyaT** [comment]
Same problem

**Comment by gagolucasm** [comment]
Same issue

**Comment by willingc** [comment]
@gnestor Would you mind taking a look and see if you can reproduce? If you can, would you mind moving to this issue to the notebook repo? Thanks.

**Comment by gnestor** [comment]
My guess: When a notebook is closed and reopened, the kernel may continue running but the cells' parent header values change. Therefore the notebook reconnects to the kernel and resumes receiving messages from the kernel but the cell outputs don't update because they don't recognize the parent headers in these messages. 

@rgbkrk Do you concur? 

The simple test case:

```py
import time

for i in range(101):
    time.sleep(1)
    print(i)
```

Then reload the notebook and if you're logging messages, you'll see that the `stream` messages continue to come in but they're not being appended to any cell's output.

What I observe:

* The cell ids change after reloading or closing/reopening (e.g. `Jupyter.notebook.get_cells()`)
* The `msg.parent_header.msg_id` is consistent for `stream` messages originating from the same cell, so even after the notebook is reloaded or closed/reopened, the parent_id is the same

I'm not sure where in the source messages are mapped to cells outputs (and specifically where the parent_id is used to do that), but I assume it has something to do with the cell's id which changes after a reload or close/reopen.

**Comment by rgbkrk** [comment]
Yup, there's no more association between the cells and the message IDs when the page is refreshed because that's all stored client side in the browser.

**Comment by cahya-wirawan** [comment]
I have also the same  problem, it's annoying if we can't see the progress anymore after we close the tab or disconnected for a while, especially during the training which could run for many hours or days. So, fix to this bug would be really very helpful. Thanks.

**Comment by jrubinstein** [comment]
please please please fix this issue! I end up having to re-run code because the output doesn't update if I have to disconnect. This is a huge problem for scripts that take a long time ... which they do. 
Using R, Chrome, on Mac

**Comment by Xpitfire** [comment]
Any update on this issue? It is impractical to operate this way on remote platforms without output updates after closing a notebook. For example Google Cloud engines runs Jupyter notebooks and the connection gets lost from time to time, but the kernel continues operating and without progress updates.

**Comment by gnestor** [comment]
@rbkrk @minrk @ellisonbg What changes would be needed to preserve the mapping between message parent header IDs and cell IDs? Could this somehow be derived from the notebook document? Or could it be stored in local storage?

**Comment by mehditlili** [comment]
I didn't have this problem before (feb 2019) but now I have it somehow. Training for 2 days and no way to see the current status..

**Comment by tausif-0311** [comment]
Any solution for this?

**Comment by gnestor** [comment]
Unfortunately no, the cell IDs change whenever the page is refreshed so the mapping between kernel messages and cells is lost. I don't have time to work on this at the moment but I think it's an important feature and I will support anyone that wants to take it on.

**Comment by dineshdaultani** [comment]
Same issue. It's really frustrating whenever the notebook closes or disconnected, the cells are not updated anymore. 
Could you please add this feature soon!

**Comment by AvivSham** [comment]
I face the same issue it is a major problem when you run a script which takes several days to run using a server. Sometimes I didn't have a choice and run the script from the beginning all over again. Can you @willingc please address that? The first issue reported in 2015 we are in 2019 and I didn't see any answer. Take for example google colab when you refresh the page still the running cell keeps updating.

**Comment by willingc** [comment]
@AvivSham Please open an issue on the [jupyter/notebook](https://github.com/jupyter/notebook) repo or the jupyterlab/jupyterlab repo with details of your particular configuration. Thanks.

**Comment by anonymone** [comment]
Pls, add the feature.

**Comment by adeperio** [comment]
Yea this bug is killing me

**Comment by harshit4084** [comment]
This feature is needed.

**Comment by henrygatech** [comment]
We need this feature!!!!

**Comment by esemeniuc** [comment]
Still exists on current jupyter

**Comment by simonsayswork** [comment]
Exists in Jupyter Lab. Makes working with it really hard. The simple way out is to transfer the code into a Python file and to execute it on the shell.

**Comment by lokeshkumarn** [comment]
Still the issue exists, I dont know when this will be solved

**Comment by jubick1337** [comment]
Yes. it is still here

**Comment by nilunima** [comment]
If I only run one cell and loose the connection to the server while the cell runs, reconnect to server (keep notebook open), cell has finished in the meantime while I lost connection, and I see the output. BUT if I run more than one cell and loose the connection to the remote server, the cells will also finish but the output will not show ("*" still in there like it is running, but it is not). Anyone else observed this?

**Comment by naono-git** [comment]
Same problem. My process will require about one week computation but I can not see the progress no longer and will not know when it finishes. Does anyone working on this issue?

**Comment by mrtpk** [comment]
I'm facing this issue. Updated everything but no avail.

**Comment by idanrichman** [comment]
same here. also facing this exact issue. thanks

**Comment by jvican** [comment]
same here. this is really annoying, please do something

**Comment by Hoeze** [comment]
Could somebody mark this issue as a priority bug?
It should really be resolved...

**Comment by sdll** [comment]
@gnestor, what are your ideas on how to resolve this?

**Comment by gnestor** [comment]
It sounds like it would require some significant refactoring and it sounds like no one is available to do it...

**Comment by sdll** [comment]
@gnestor, this is sad. Could you please expand on some specific parts to refactor?

**Comment by regorigregory** [comment]
Perhaps save STD out to a {file}-{cell}-{hash?} format and create a speparete notebook which you can use to view any notebook's any cell's output dynamically updated (as it might be still running)?

I know it is not ideal, but could write it in a jiffy...and would resolve the lost output problem.

**Comment by Ivsucram** [comment]
The issue still exists, why is it marked as closed?

**Comment by dineshdaultani** [comment]
> It's been more than 4 years since this issue has been reported, yet still no update.  Although I feel like it should be treated as must-have core feature of notebook rather than nice-to-have.

**Comment by marc9323** [comment]
Same problem.  If I open a notebook the first time I get the output.  But if I re open or try to run a second time there is no output.  I have had nothing but problems with Jupyter notebooks.  Yes, they are a great way of sharing data but it's simply too much trouble.

**Comment by pharmhax** [comment]
Same issue here. Very unfortunate for long-running calculations.

**Comment by solyoh21** [comment]
Thank you
Great job!!!

Em ter, 28 de abr de 2020 17:03, Matt Maciejewski <notifications@github.com>
escreveu:

**Comment by Hoeze** [comment]
I think we have to wait for https://github.com/jupyterlab/rtc/ to fix this issue.

**Comment by mcmar** [comment]
For anyone looking for a quick workaround:
- Open the Chrome dev tools
- Go to network tab
- Click `WS` (Websockets)
- Click on any of the active websockets
- Go to messages tab
- Click on any of the messages after the first
- Expand content
You should see your output there

**Comment by ruthvik92** [comment]
> For anyone looking for a quick workaround:
> 
>     * Open the Chrome dev tools
> 
>     * Go to network tab
> 
>     * Click `WS` (Websockets)
> 
>     * Click on any of the active websockets
> 
>     * Go to messages tab
> 
>     * Click on any of the messages after the first
> 
>     * Expand content
>       You should see your output there

Do you have similar solution for firefox?

**Comment by sambbhavgarg** [comment]
Thanks @mcmar! Works on MS Edge as well. Any idea why the jupyter notebook doesnt write these updates? There are a few errors on my console, do you think they're the reason?
![Screenshot (85)_LI](https://user-images.githubusercontent.com/32191060/88569304-e02d0380-d057-11ea-9b48-82b8fa1aef7a.jpg)

**Comment by sambbhavgarg** [comment]
Possible workaround for people using Keras on jupyter notebook, save your model weights. If by any chance you refresh the window, wait for the saved weights to update, and load freshly saved models, compile and start training again.

**Comment by HJJ256** [comment]
Another workaround can be to create a custom callback for keras which saves your metrics and losses to a CSV file, in that way, the CSV will keep updating after every epoch as the model trains.

**Comment by HJJ256** [comment]
> Possible workaround for people using Keras on jupyter notebook, save your model weights using keras callbacks after every N epochs.. If by any chance you refresh the window, wait for the saved weights to update, and load freshly saved models, compile and start training again.
> 
> ```
> checkpoint_filepath = 'ckpts'
> fitsize = int(batch_size*((len(X_train)*0.8)//batch_size))
> model_checkpoint_callback = ModelCheckpoint(
>     filepath=checkpoint_filepath,
>     save_weights_only=True,
>     monitor='val_acc',
>     mode='max',
>     verbose=1,
>     save_best_only=True)
> history = model.fit(X_train[:fitsize], 
>                     y_train[:fitsize], 
>                     batch_size=batch_size, 
>                     epochs=50, 
>                     verbose=1, 
>                     validation_data=(X_train[fitsize:], 
>                                      y_train[fitsize:]),
>                    callbacks=[model_checkpoint_callback])
> ```

This might ruin your random states though, and botch the experiment in the process.

**Comment by sambbhavgarg** [comment]
> Another workaround can be to create a custom callback for keras which saves your metrics and losses to a CSV file, in that way, the CSV will keep updating after every epoch as the model trains.

That's what Im doing now!

**Comment by darkmatter08** [comment]
Still an issue! This is critical when working with remote environments like AWS with long running cells, for example, for ML.

**Comment by sambbhavgarg** [comment]
> Still an issue! This is critical when working with remote environments like AWS with long running cells, for example, for ML.

Hey @darkmatter08 
Have you tried using this: https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/CSVLogger?

**Comment by darkmatter08** [comment]
@sambbhavgarg I'm using PyTorch -- by chance are you aware of an equivalent there? 
My simple workaround is to setup a python logger that logs to stdout and to a file. However, I lose out on the ability to display images inline in the notebook.

**Comment by bixing-udemy** [comment]
Same issue. It would be good if this issue can be fixed.

**Comment by rgbkrk** [comment]
Here's the repository to follow for the effort: https://github.com/jupyterlab/rtc

**Comment by sebtackin** [comment]
just to add, this is not a Keras' issue but Jupiter's one. If you are using AWS, than AWS adds its 5 cents too. If your job runs more than 12h the notebook logs-out (and you need to re-login) and you run into the above. I tried to ask for longer period before logouts but no success. but the kernel runs, your calculations progress. to see it you can make your script write sth to the drive at the end of some iteration. than you will see changing message in the file system "file has been updates seconds, x minutes,... ago". in my case it flips between. minutes and seconds constantly and thus I know it works also make sure you have auto shutdown script on (in AWS). but indeed VEEERRYYYY ANNOOYYIINGGGG!!!!

**Comment by eduOS** [comment]
I'd like to rephrase this question to how I can redirect the output of Jupyter Notebook to a file?

**Comment by alex-lt-kong** [comment]
Same issue. It would be good if this issue can be fixed.

**Comment by sanxchep** [comment]
schrodinger's kernel lmao

Whether the process is running as intended or not

**Comment by zer0-memory** [comment]
I have this issue even when running Jupyter/JupyterLab with --collaborative; is there any hope for a fix?

**Comment by apayne97** [comment]
here to keep the issue alive :). same problem. guess they will never fix it lol.

**Comment by alexgarel** [comment]
If someone propose a PR, I think they will accept it, indeed !

**Comment by halqadasi** [comment]
We are in 2022, the problem is still unsolved :(

**Comment by sanxchep** [comment]
happened again when i was running a model....
why is this a thing in 2022 :(

**Comment by ionutscorta** [comment]
Just use the following `jupyter nbconvert <notebook> --to ipynb --execute --debug` in a terminal with tmux active. This will run in the background and plot everything you have in the notebook. The downside is that the notebook is closed when it's done but the good side is you get the plots and metrics from your notebook and if you write it correctly you get the checkpoints from the model.

Annoying issue but who knows when it will be fixed.

**Comment by FrankFundel** [comment]
You can also use weights and biases for logging.

**Comment by nershman** [comment]
There's a simple workaround using ipython `%%capture` magic.

```
%%capture cap
#Whatever you want to run goes here

#save it to a file
with open('cell_output.txt', 'w') as f:
    f.write(cap.stdout)
```
When you re-open your notebook you can either load the text file or just print the variable you saved output to:
<img width="602"  src="https://user-images.githubusercontent.com/755742/167730147-0cf2e5b1-55f2-485f-b603-d9931dfaac1f.png">

Not ideal since it doesn't display the contents as it runs, but better than nothing I suppose.

**Comment by talha1503** [comment]
This problem is still unsolved :/ It's difficult to use jupyter with remote environments, having to keep it open all the time.

**Comment by SoftFeta** [comment]
This bug happens all the time.

**Comment by tommedema** [comment]
Is this still being considered?

**Comment by YogiOnBioinformatics** [comment]
PLEASE CREATE A SOLUTION FOR THIS!!!!!!!!!!!

**Comment by MichalRIcar** [comment]
Hello, same beg from our side - remapping jupyter server and open close jupyter notebook cause 0 continuations..thus for the real-world application NTB is unusable, yet, it is a great "app", it would be great to add this flexible/dynamical element if possible. Or at least a workaround to it.

**Comment by rgbkrk** [comment]
I've been working on a platform to solve exactly this, called [Noteable](https://noteable.io/). You can try it yourself at https://app.noteable.io/

The notebook outputs are updated even if your tabs are closed. You can also open multiple tabs and work on the same document with yourself or others. No worries about clobbering your work and you can let long-running jobs run. Reach out if you'd like to get wider support at your lab or organization.

**Comment by c0d3rman** [comment]
Here we are, 7+ years later, still literally zero progress on this. Every single person who's ever used a Jupyter notebook for ML has run into this problem and many lost major amounts of progress because of it. We are not going to stop asking for this until you fix it. It's not a nice-to-have, it's the biggest problem in Jupyter notebooks today, and it needs to be fixed. Does the dev team just not care?

**Comment by sanxchep** [comment]
Is this application even maintained nowadays?

**Comment by axel588** [comment]
same issue 7 years later, what a shame

**Comment by chunqiaohu** [comment]
您好，您的邮件我已经收到，会尽快给您回复，谢谢。

**Comment by apayne97** [comment]
> same issue 7 years later, what a shame

I kind of love the occasionally messages I get from this thread though

**Comment by thoangtrvn** [comment]
same issue for me, when running the notebook with port-forwarding.

**Comment by sambbhavgarg** [comment]
> > same issue 7 years later, what a shame
> 
> I kind of love the occasionally messages I get from this thread though

frfr, if the devs have not taken it up for 7y, theyre probs not gonna do it now. The best solution so far is to simply update a text/csv with progress every epoch/iteration, and move on...

**Comment by xnohat** [comment]
> For anyone looking for a quick workaround:
> 
> * Open the Chrome dev tools
> * Go to network tab
> * Click `WS` (Websockets)
> * Click on any of the active websockets
> * Go to messages tab
> * Click on any of the messages after the first
> * Expand content
>   You should see your output there

another beautified output workaround . Use workaround above with this script in chrome dev tool console

```
//copy current websocket and change session_id by padding ...-2
const socket = new WebSocket('ws://jupyterdomain.com/api/kernels/c0d04b6d-4a6b-4e21-9e5a-52d4cd7c93f7/channels?session_id=a6ac43c6-4211-4699-b1da-43af501df1b1-2');

// Listen for messages
socket.addEventListener('message', (event) => {
    //console.log('Message from server ', event.data)
    console.log(JSON.parse(event.data).content.text)
});

```

**Comment by thoangtrvn** [comment]
I also found a workaround solution. Just don't refresh the browser tab. If you're using Jupyter lab, there are multiple panes,
just close the pane for that notebook, and then on the left side, click to reopen the notebook again.

**Comment by 362227** [comment]
7 years lol

**Comment by jfs42** [comment]
super lame this wasn't addressed, guess it's time to stop using Jupyter lab...

**Comment by ivanov** [comment]
Hello, fellow humans,

  There are 68 participants here, but I am going to close this issue, since it has long ceased to be productive. There are some good gems sprinkled about, like [the mention of the `%%capture` magic](https://github.com/jupyter/jupyter/issues/83#issuecomment  -1122946511), but the tone of most messages on this issue come from a misplaced sense of entitlement, something to the effect of:

  > What a bunch of jokers these Jupyter developers, they made and continue to work on this tool we benefit from, make it available for free with a permissive license, but when we tell them about a known limitation of how it works, without contributing anything other than periodic "we still want this" message, they don't do anything about it.

~ Trust me - if it was easy within the current framework within Jupyter, we would have done it by now. And if you don't trust me, no one has stopped you from trying to do it yourself. But piling on with a "bump!" or "PLEASE FIX THIS!" doesn't do anything
  (except maybe chip away at my faith in the human spirit). Think about this: is it more sad that this situation is perceived to be the same as it stood eight years ago, or that so many people keep coming here to demand their outputs of long running notebooks without pitching in themselves?

  I also worked on the Noteable platform.  And no, @sambbhavgarg, though he does have superhuman hacking powers,  @rgbkrk didn't create it by himself :) , there were about a dozen people working on it, and no, it wasn't created just to solve this one issue. Importantly, the way it did get solved there involved a significant engineering effort to move the state of the notebook from the browser to the server, which is one of the approaches that's also been happening in parallel out in the open source community.


  Those of you wanting to follow along and interested in seeing this addressed in Jupyter are encouraged to follow these issues, or use them as starting points to getting involved.

  https://github.com/jupyter/notebook/issues/641
  https://github.com/jupyterlab/jupyterlab/issues/2833
  https://github.com/jupyter-server/jupyter_server/issues/1274
  https://github.com/jupyter-server/message_replay

  I'll be closing this issue, but if you want to continue the discussion, you can do so over on the Jupyter Discourse.
Happy hacking! :bowtie:
